integration_test/integration-test.go

```diff
@@ -0,0 +1,313 @@
+package main
+
+import (
+	"io/ioutil"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+	"strings"
+	"os"
+)
+
+/*****************************
+* INTEGRATION TEST FRAMEWORK *
+******************************/
+
+type Map = map[string]string
+
+/*
+* TestSuite struct
+*   keep track of info used by a test suite
+*   different proxy configurations corresponds
+*   to different test suites
+*/
+type TestSuite struct {
+	// url of the proxy to be tested
+	// the test client will send requests to this url
+	proxy_url string
+
+	// the port that the test server will be listening at
+	server_port string
+
+	// the expected message received from the test server if
+	// the request has been successful
+	server_message string
+
+	// the server that will be responding requests from the proxy
+	server http.Server
+
+	// the channels for communication between the server
+	// and the client. Received requests will be pushed through these
+	// channels by the server for the client to verify
+	request_received chan *http.Request
+	body_string_received chan string
+
+	// a collection of requests to run in this suite
+	// each test has a test name associated with it
+	requests map[string]TestRequest
+}
+
+/*
+* TestRequest struct
+*   keep track of info used by one request in a test suite
+*/
+type TestRequest struct {
+	// the test request to be made
+	path string
+	pathParams url.Values
+	headers Map
+	cookies Map
+	body  url.Values
+
+	// the expected sqli detection result for the proxy
+	// if true, the request contains sqli
+	sqli bool
+	sqli_body bool
+}
+
+/*
+* Initiate a new test suite with a given proxy_url
+*/
+func NewTestSuite(proxy_url string) TestSuite {
+	return TestSuite{
+		proxy_url: proxy_url,
+		server_port: ":8080",
+		server_message: "request received",
+		request_received: make(chan *http.Request),
+		body_string_received: make(chan string),
+		requests: map[string]TestRequest{},
+	}
+}
+
+/*
+* Run all requests in a test suite
+*/
+func (t TestSuite) RunRequests() {
+	t.StartServer()
+	t.CheckProxyConnection()
+	for name, test := range t.requests {
+		t.RunTestRequest(name, test)
+	}
+	t.CloseServer()
+}
+
+/*
+* Initiate a new test request with default content
+*/
+func NewTestRequest() TestRequest {
+	return TestRequest{
+		path: "/",
+		pathParams: url.Values{"path-key-0": {"path-val-0"}},
+		headers: Map{"header-key-0": "header-val-0"},
+		cookies: Map{"cookie-key-0": "cookie-val-0"},
+		body: url.Values{"body-key-0": {"body-val-0"}},
+		sqli: false,
+		sqli_body: false,
+	}
+}
+
+/*
+* Run a given test in a test suite
+* assuming the server and the proxy has been running
+*/
+func (t TestSuite) RunTestRequest(test_name string, test TestRequest) {
+	result := t.Eval(test)
+	if !result {
+		fmt.Println("TEST FAILED: " + test_name)
+		fail()
+	}
+	fmt.Println("TEST PASSED: " + test_name)
+}
+
+/*
+* Initialize and Start the test server
+*/
+func (suite TestSuite) StartServer() {
```

Comment:
receiver name `suite` should be consistent with previous receiver name `t` for `TestSuite`